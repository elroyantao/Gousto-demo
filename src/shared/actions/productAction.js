import 'isomorphic-fetch'

export const fetchProductsRequest = () => ({
  type: 'FETCH_PRODUCTS_REQUEST'
})

export const fetchProductsResponse = (products) => ({
  type: 'FETCH_PRODUCTS_RESPONSE',
  products
})

export const fetchProductsFailure = () => ({
  type: 'FETCH_PRODUCTS_FAILURE'
})

export const fetchProducts = () => (dispatch) => {
  dispatch(fetchProductsRequest())
  return fetch('/api/products', { method: 'GET' })
  .then((res) => {
    if (!res.ok) throw Error(res.statusText)
    return res.json()
  })
  .then((res) => {
    dispatch(fetchProductsResponse(res.data))
  })
  .catch(() => {
    dispatch(fetchProductsFailure())
  })
}
