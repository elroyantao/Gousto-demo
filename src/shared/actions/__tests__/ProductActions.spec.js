import fetchMock from 'fetch-mock'
import * as actions from '../productAction'

const fetchAction = actions.fetchProducts()
const dispatch = jest.fn()

describe('product actions', () => {
  it('fetchProductsRequest()', () => {
    expect(actions.fetchProductsRequest()).toMatchSnapshot()
  })
  it('fetchProductsFailure()', () => {
    expect(actions.fetchProductsFailure()).toMatchSnapshot()
  })
  it('fetchProductsResponse()', () => {
    expect(actions.fetchProductsResponse([1, 2, 3])).toMatchSnapshot()
  })
  describe('fetchProducts() ', () => {
    // fetch.mockResponse(JSON.stringify({ access_token: '12345' }))
    afterEach(() => {
      fetchMock.restore()
      jest.clearAllMocks()
    })
    it('a successful fetch', () => {
      fetchMock.get('/api/products', {
        data: [1, 2, 3]
      })
      expect.assertions(3)
      return fetchAction(dispatch).then(() => {
        expect(fetchMock.lastCall()).toEqual(['/api/products', { method: 'GET' }])
        expect(dispatch).toHaveBeenCalledTimes(2)
        expect(dispatch.mock.calls).toMatchSnapshot()
      })
    })
    it('an unsuccessful fetch', () => {
      fetchMock.get('/api/products', 402)
      expect.assertions(3)
      return fetchAction(dispatch).then(() => {
        expect(fetchMock.lastCall()).toEqual(['/api/products', { method: 'GET' }])
        expect(dispatch).toHaveBeenCalledTimes(2)
        expect(dispatch.mock.calls).toMatchSnapshot()
      })
    })
  })
})
