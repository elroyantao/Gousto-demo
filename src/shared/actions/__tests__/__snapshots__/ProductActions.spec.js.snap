// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`product actions fetchProducts()  a successful fetch 1`] = `
Array [
  Array [
    Object {
      "type": "FETCH_PRODUCTS_REQUEST",
    },
  ],
  Array [
    Object {
      "products": Array [
        1,
        2,
        3,
      ],
      "type": "FETCH_PRODUCTS_RESPONSE",
    },
  ],
]
`;

exports[`product actions fetchProducts()  an unsuccessful fetch 1`] = `
Array [
  Array [
    Object {
      "type": "FETCH_PRODUCTS_REQUEST",
    },
  ],
  Array [
    Object {
      "type": "FETCH_PRODUCTS_FAILURE",
    },
  ],
]
`;

exports[`product actions fetchProductsFailure() 1`] = `
Object {
  "type": "FETCH_PRODUCTS_FAILURE",
}
`;

exports[`product actions fetchProductsRequest() 1`] = `
Object {
  "type": "FETCH_PRODUCTS_REQUEST",
}
`;

exports[`product actions fetchProductsResponse() 1`] = `
Object {
  "products": Array [
    1,
    2,
    3,
  ],
  "type": "FETCH_PRODUCTS_RESPONSE",
}
`;
